### PR TITLE
⚠ Cache's ByObject is now a map Object -> options

### PR DESCRIPTION
**Pull request by vincepri**
This is a breaking change and revisits some of the previous changes we've made. It brings the options ByObject to a top level map, which now contains all the options the internal informer can be configured with. In addition, it removes the ObjectAll object, which was super confusing and only used to disable the deep copy for every object; instead it's now possible to disable deep copy through a top level option.

Signed-off-by: Vince Prignano <vincepri@redhat.com>

### CODE REVIEW

File: pkg/cache/cache.go [hunk 1]
```diff
@@ -147,31 +134,36 @@ type ViewOptions struct {
...
 	// and for updated objects.
-	Transform TransformByObject
+	Transform toolscache.TransformFunc
 
 	// UnsafeDisableDeepCopy indicates not to deep copy objects during get or
 	// list objects per GVK at the specified object.
 	// Be very careful with this, when enabled you must DeepCopy any object before mutating it,
 	// otherwise you will mutate the object in the cache.
-	UnsafeDisableDeepCopy DisableDeepCopyByObject
+	UnsafeDisableDeepCopy bool
```
Comment (varshaprasad96) [inline pkg/cache/cache.go:166]: @vincepri just curious, is it intentional to not overwrite `UnsafeDisableDeepCopy` by the top level `opts. UnsafeDisableDeepCopy` if it's not specified per object? (I understand its safe to default it to false, but just asking if it would be confusing to the user). Also is there a reason to have it as a pointer in top level and not in byObject?

Comment (vincepri) [reply pkg/cache/cache.go:166]: The thinking behind this was to retain some of the features we have today, and make it a bit safer as you said. 

- If a user wants to disable deep copy globally, we should be able to override it for specific objects.
- If users want to keep it to false globally (default), we should be able to override it to true for specific objects.

The reason for the top level pointer is for merging options together, there is a need to know if the option was left unset, or was false, which is the default of a boolean field. The option inheritance logic is hard to follow (spent a good deal of time on it today 😄), but it basically requires the pointer to know if we should inherit or not.

Happy to chat about it more if we feel like this should be made safer or easier to use. I would expect though that most users shouldn't disable deep copy unless they absolutely know the repercussions of doing so, which is an advanced use case

File: pkg/cache/cache.go [hunk 2]
```diff
@@ -147,31 +134,36 @@ type ViewOptions struct {
 	DefaultTransform toolscache.TransformFunc
 
 	// ByObject restricts the cache's ListWatch to the desired fields per GVK at the specified object.
-	ByObject ViewByObject
+	ByObject map[client.Object]ByObject
+
+	// UnsafeDisableDeepCopy indicates not to deep copy objects during get or
+	// list objects for EVERY object.
+	// Be very careful with this, when enabled you must DeepCopy any object before mutating it,
+	// otherwise you will mutate the object in the cache.
```
Comment (alvaroaleman) [inline pkg/cache/cache.go:142]: Nit: mention that this can be overrriden through `ByObject` per GVK?

**Review by alvaroaleman — COMMENTED** [review]
On

File: pkg/cache/cache.go (hunk 1 above)
Comment (varshaprasad96) [reply pkg/cache/cache.go:166]: Ah, I see, got it! Thanks for explaining @vincepri :)

File: pkg/cache/cache.go [hunk 3]
```diff
@@ -147,31 +134,36 @@ type ViewOptions struct {
...
-	// Selectors restricts the cache's ListWatch to the desired
-	// fields per GVK at the specified object, the map's value must implement
-	// Selectors [1] using for example a Set [2]
-	// [1] https://pkg.go.dev/k8s.io/apimachinery/pkg/fields#Selectors
-	// [2] https://pkg.go.dev/k8s.io/apimachinery/pkg/fields#Set
-	Selectors SelectorsByObject
+// ByObject offers more fine-grained control over the cache's ListWatch by object.
+type ByObject struct {
+	// Label represents a label selector for the object.
+	Label labels.Selector
```
[the remark this answers is not in the document]
Comment (vincepri) [reply pkg/cache/cache.go:149]: This was a suggestion by @alvaroaleman, I'm happy to do either way, although to reduce verbosity of the ByObject struct, might make sense to keep it this way

File: pkg/cache/cache.go [hunk 4]
```diff
@@ -147,31 +134,36 @@ type ViewOptions struct {
...
-// ViewByObject offers more fine-grained control over the cache's ListWatch by object.
-type ViewByObject struct {
-	// Selectors restricts the cache's ListWatch to the desired
-	// fields per GVK at the specified object, the map's value must implement
-	// Selectors [1] using for example a Set [2]
-	// [1] https://pkg.go.dev/k8s.io/apimachinery/pkg/fields#Selectors
-	// [2] https://pkg.go.dev/k8s.io/apimachinery/pkg/fields#Set
-	Selectors SelectorsByObject
+// ByObject offers more fine-grained control over the cache's ListWatch by object.
+type ByObject struct {
```
[the remark this answers is not in the document]
Comment (vincepri) [reply pkg/cache/cache.go:147]: Can rename if we want to, the Struct is only used within the options, and was trying to reduce overall verbosity

[the remark this answers is not in the document]
Comment (alvaroaleman) [reply pkg/cache/cache.go:147]: Maybe `ObjectOptions`? But no strong opining either way

File: pkg/cache/cache.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alvaroaleman) [reply pkg/cache/cache.go:149]: yeah IMHO this is easier to use when keeping them separate. I guess we could also do the same for the `DefaultSelector`

File: pkg/cache/cache.go (hunk 2 above)
Comment (vincepri) [reply pkg/cache/cache.go:142]: fixed, thanks

File: pkg/cache/cache.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (vincepri) [reply pkg/cache/cache.go:151]: As in, separate `DefaultLabelSelector` and `DefaultFieldSelector`?

File: pkg/cache/cache.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (vincepri) [reply pkg/cache/cache.go:149]:
```go
cache.Options{
  ByObject: map[client.Object]cache.ByObject{
    &corev1.Pod{}: {}
  }
}
```

IMO, this reads well and looks nice. @FillZpp if you feel that we should change it, happy to do it though

File: pkg/cache/cache_unit_test.go [hunk 5]
```diff
@@ -110,120 +109,131 @@ var _ = Describe("cache.inheritFrom", func() {
...
+			specified.ByObject = map[client.Object]ByObject{
+				&corev1.Pod{}: {},
+			}
+			Expect(checkError(specified.inheritFrom(inherited)).ByObject).To(HaveLen(1))
 		})
 		It("is inherited when only inherited is set", func() {
 			inherited.Scheme = coreScheme
-			inherited.View.ByObject.Selectors = map[client.Object]ObjectSelector{&corev1.ConfigMap{}: {}}
-			Expect(checkError(specified.inheritFrom(inherited)).View.ByObject.Selectors).To(HaveLen(1))
+			specified.ByObject = map[client.Object]ByObject{
```
Comment (sbueringer) [inline pkg/cache/cache_unit_test.go:142]: should be inherited

File: pkg/cache/cache_unit_test.go [hunk 6]
```diff
@@ -110,120 +109,131 @@ var _ = Describe("cache.inheritFrom", func() {
...
+			inherited.ByObject = map[client.Object]ByObject{
+				&corev1.ConfigMap{}: {},
+			}
+			Expect(checkError(specified.inheritFrom(inherited)).ByObject).To(HaveLen(1))
 		})
 		It("uses inherited scheme for specified selectors", func() {
 			inherited.Scheme = coreScheme
-			specified.View.ByObject.Selectors = map[client.Object]ObjectSelector{&corev1.ConfigMap{}: {}}
-			Expect(checkError(specified.inheritFrom(inherited)).View.ByObject.Selectors).To(HaveLen(1))
+			inherited.ByObject = map[client.Object]ByObject{
```
Comment (sbueringer) [inline pkg/cache/cache_unit_test.go:200]: should be specified

File: pkg/cache/cache_unit_test.go [hunk 7]
```diff
@@ -110,120 +109,131 @@ var _ = Describe("cache.inheritFrom", func() {
...
+				&corev1.ConfigMap{}: {},
+			}
+			Expect(checkError(specified.inheritFrom(inherited)).ByObject).To(HaveLen(1))
 		})
 		It("uses specified scheme for specified selectors", func() {
 			specified.Scheme = coreScheme
-			specified.View.ByObject.Selectors = map[client.Object]ObjectSelector{&corev1.ConfigMap{}: {}}
-			Expect(checkError(specified.inheritFrom(inherited)).View.ByObject.Selectors).To(HaveLen(1))
+			inherited.ByObject = map[client.Object]ByObject{
```
Comment (sbueringer) [inline pkg/cache/cache_unit_test.go:207]: same

File: pkg/cache/cache.go [hunk 8]
```diff
@@ -147,31 +134,38 @@ type ViewOptions struct {
...
 	// and for updated objects.
-	Transform TransformByObject
+	Transform toolscache.TransformFunc
 
 	// UnsafeDisableDeepCopy indicates not to deep copy objects during get or
 	// list objects per GVK at the specified object.
 	// Be very careful with this, when enabled you must DeepCopy any object before mutating it,
 	// otherwise you will mutate the object in the cache.
-	UnsafeDisableDeepCopy DisableDeepCopyByObject
+	UnsafeDisableDeepCopy bool
```
[the remark this answers is not in the document]
Comment (vincepri) [reply pkg/cache/cache.go:168]: I was thinking the latter, if we're overriding something I _guess_ that would be the source of truth. That said, the bool's default to false comes into play into unexpected UX. Should we convert to pointer here as well?

File: pkg/cache/cache.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alvaroaleman) [reply pkg/cache/cache.go:151]: Yeah

File: pkg/cache/cache.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply pkg/cache/cache.go]: I think pointer gives the better UX

File: pkg/cache/cache.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (vincepri) [reply pkg/cache/cache.go:154]: done, ptal

File: pkg/cache/cache_test.go [hunk 9]
```diff
@@ -1822,12 +1810,11 @@ func isKubeService(svc metav1.Object) bool {
...
 
 func isPodDisableDeepCopy(opts cache.Options) bool {
-	if d, ok := opts.View.ByObject.UnsafeDisableDeepCopy[&corev1.Pod{}]; ok {
-		return d
-	} else if d, ok = opts.View.ByObject.UnsafeDisableDeepCopy[cache.ObjectAll{}]; ok {
-		return d
-	} else if d, ok = opts.View.ByObject.UnsafeDisableDeepCopy[&cache.ObjectAll{}]; ok {
-		return d
+	if opts.ByObject[&corev1.Pod{}].UnsafeDisableDeepCopy != nil && *opts.ByObject[&corev1.Pod{}].UnsafeDisableDeepCopy {
+		return true
```
Comment (sbueringer) [inline pkg/cache/cache_test.go:1814]: Shouldn't this return the value of UnsafeDisableDeepCopy?

(same below)

Comment (sbueringer) [reply pkg/cache/cache_test.go:1814]: I.e. if it's set to false on the object shouldn't it be false?

Comment (vincepri) [reply pkg/cache/cache_test.go]: Changed it, there shouldn't be any logical difference, but I agree it might be more clear

Comment (vincepri) [reply pkg/cache/cache_test.go]: Actually I stand corrected, we want to always return the most specific value now, given that's a pointer

Comment (sbueringer) [reply pkg/cache/cache_test.go]: Yup, that's what I thought :)

**Comment by sbueringer** [comment]
lgtm from my side

**Review by alvaroaleman — APPROVED** [review]
